tools/tpm2_key_import.c

```diff
@@ -0,0 +1,578 @@
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//**********************************************************************;
+
+#include <stdbool.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+
+#include <openssl/rand.h>
+#include <openssl/bn.h>
+#include <openssl/evp.h>
+#include <openssl/sha.h>
+#include <openssl/hmac.h>
+#include <openssl/aes.h>
+#include <openssl/rsa.h>
+
+#include <getopt.h>
+#include <limits.h>
+#include <sapi/tpm20.h>
+#include <sapi//marshal.h>
+
+#include "log.h"
+#include "files.h"
+#include "tpm_kdfa.h"
+#include "main.h"
+#include "options.h"
+#include "tpm2_util.h"
+
+#define SYM_KEY_SIZE 16
+#define max_buffer_size  1024
+
+typedef struct tpm_import_ctx tpm_import_ctx;
+struct tpm_import_ctx {
+    char *input_key_file;
+    char *import_key_public_file;
+    char *import_key_private_file;
+    char *parent_key_public_file;
+    uint8_t *input_key_buffer;
+    //Parent public key for seed encryption
+    TPM2B_PUBLIC_KEY_RSA parent_public_key;
+    TPM_HANDLE parent_key_handle;
+    //External key public
+    TPM2B_PUBLIC import_key_public;
+    //External key name
+    TPM2B_NAME import_key_public_name;
+    //External key sensitive
+    TPM2B_SENSITIVE import_key_sensitive;
+    //External key private
+    TPM2B_PRIVATE import_key_private;
+    //Protection Seed and keys
+    uint8_t protection_seed_data[SHA256_DIGEST_SIZE]; //max tpm digest
+    uint8_t encrypted_protection_seed_data[MAX_RSA_KEY_BYTES];
+    uint8_t protection_hmac_key[SHA256_DIGEST_SIZE];
+    uint8_t protection_enc_key[SYM_KEY_SIZE];
+    uint8_t import_key_public_unique_data[SHA256_DIGEST_SIZE];
+    uint8_t outer_integrity_hmac[SHA256_DIGEST_SIZE];
+    TPM2B_DATA enc_sensitive_key;
+    TPM2B_MAX_BUFFER encrypted_inner_integrity;
+    TPM2B_MAX_BUFFER encrypted_duplicate_sensitive;
+    //SAPI and TCTI context
+    TSS2_SYS_CONTEXT *sapi_context;
+};
+
+int ssl_RSA_set0_key(RSA *r, BIGNUM *n, BIGNUM *e, BIGNUM *d) {
+
+    if ((r->n == NULL && n == NULL) || (r->e == NULL && e == NULL))
+        return 0;
+    if (n != NULL) {
+        BN_free(r->n);
+        r->n = n;
+    }
+    if (e != NULL) {
+        BN_free(r->e);
+        r->e = e;
+    }
+    if (d != NULL) {
+        BN_free(r->d);
+        r->d = d;
+    }
+    return 1;
+}
+
+bool encrypt_seed_with_tpm2_rsa_public_key(tpm_import_ctx *ctx) {
```

Comment:
Comments in this routine that explain at a high level what your doing would be super helpful.

Reply:
A function this big and complicated should have a comment header explaining it's purpose.